In [2]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding
encoder = OneHotEncoder(sparse_output=False)
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Bangun model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(8, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Kompilasi
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model
model.fit(X_train, y_train, epochs=50, batch_size=8)

# Evaluasi
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi: {acc}")


c:\Users\USER\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.3250 - loss: 1.3027   
Epoch 2/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3250 - loss: 1.1362 
Epoch 3/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.3250 - loss: 1.0863 
Epoch 4/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3250 - loss: 1.0476 
Epoch 5/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3333 - loss: 1.0270 
Epoch 6/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3250 - loss: 1.0010 
Epoch 7/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3250 - loss: 0.9815 
Epoch 8/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.3333 - loss: 0.9643 
Epoch 9/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3917 - loss: 0.9439 
Epoch 10/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6083 - loss: 0.9235 
Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6417 - loss: 0.9069 
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy

## TUGAS 2: VARIASI HIDDEN LAYERS

Ubah jumlah neuron hidden layer.

Bandingkan akurasi dengan konfigurasi awal.

In [4]:
import pandas as pd


In [5]:
# ========== TUGAS 2: VARIASI HIDDEN LAYERS ==========
print("\n" + "=" * 70)
print("TUGAS 2: VARIASI JUMLAH NEURON HIDDEN LAYER")
print("=" * 70)

# Konfigurasi yang akan diuji
configs = [
    {'name': 'Original [10, 8]', 'layers': [10, 8]},
    {'name': 'Small [5, 3]', 'layers': [5, 3]},
    {'name': 'Medium [15, 10]', 'layers': [15, 10]},
    {'name': 'Large [20, 15]', 'layers': [20, 15]},
    {'name': 'Deep [12, 10, 8]', 'layers': [12, 10, 8]},
]

results_config = []

for config in configs:
    print(f"\n--- Testing: {config['name']} ---")
    
    # Reset seed
    tf.random.set_seed(42)
    
    # Build model
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(config['layers'][0], activation='relu', input_shape=(4,)))
    
    for neurons in config['layers'][1:]:
        model.add(tf.keras.layers.Dense(neurons, activation='relu'))
    
    model.add(tf.keras.layers.Dense(3, activation='softmax'))
    
    # Compile
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    
    # Train
    history = model.fit(
        X_train, y_train,
        epochs=50,
        batch_size=8,
        validation_data=(X_test, y_test),
        verbose=0
    )
    
    # Evaluate
    loss, acc = model.evaluate(X_test, y_test, verbose=0)
    
    results_config.append({
        'Konfigurasi': config['name'],
        'Loss': loss,
        'Akurasi': acc,
        'History': history
    })
    
    print(f"Loss: {loss:.4f}, Akurasi: {acc:.4f} ({acc*100:.2f}%)")

# Tampilkan tabel perbandingan
print("\n" + "=" * 70)
print("PERBANDINGAN HASIL TUGAS 2")
print("=" * 70)
df_config = pd.DataFrame([{
    'Konfigurasi': r['Konfigurasi'],
    'Loss': f"{r['Loss']:.4f}",
    'Akurasi': f"{r['Akurasi']:.4f}",   
    'Akurasi (%)': f"{r['Akurasi']*100:.2f}%"
} for r in results_config])
print(df_config.to_string(index=False))



TUGAS 2: VARIASI JUMLAH NEURON HIDDEN LAYER

--- Testing: Original [10, 8] ---
Loss: 0.2049, Akurasi: 0.9000 (90.00%)

--- Testing: Small [5, 3] ---
Loss: 1.1001, Akurasi: 0.3000 (30.00%)

--- Testing: Medium [15, 10] ---
Loss: 0.2058, Akurasi: 0.9333 (93.33%)

--- Testing: Large [20, 15] ---
Loss: 0.2171, Akurasi: 0.9333 (93.33%)

--- Testing: Deep [12, 10, 8] ---
Loss: 0.1470, Akurasi: 0.9667 (96.67%)

PERBANDINGAN HASIL TUGAS 2
     Konfigurasi   Loss Akurasi Akurasi (%)
Original [10, 8] 0.2049  0.9000      90.00%
    Small [5, 3] 1.1001  0.3000      30.00%
 Medium [15, 10] 0.2058  0.9333      93.33%
  Large [20, 15] 0.2171  0.9333      93.33%
Deep [12, 10, 8] 0.1470  0.9667      96.67%


Model awal memiliki akurasi **0.93**. Ketika jumlah neuron terlalu sedikit (**[5, 3]**), akurasinya turun drastis menjadi **30%** karena jaringan tidak mampu belajar pola dengan baik. Pada konfigurasi yang lebih besar seperti **[15, 10]** dan **[20, 15]**, akurasi kembali setara dengan model awal. Konfigurasi **Deep [12, 10, 8]** memberikan hasil terbaik (**0.96**) karena jaringan lebih dalam sehingga mampu menangkap pola data lebih kompleks. Dengan demikian, jumlah neuron yang terlalu kecil menurunkan performa, sedangkan struktur lebih dalam dapat meningkatkan akurasi hingga batas tertentu.


## Tugas 3

Bandingkan Sigmoid vs ReLU pada dataset Iris.

Catat perbedaan loss dan akurasi.

### **Perbandingan Sigmoid dan ReLU**

Pada model yang menggunakan **Sigmoid**, proses pelatihan berjalan lebih lambat karena nilai aktivasi cenderung mengecil (gradient mengecil), sehingga jaringan membutuhkan iterasi yang lebih banyak untuk mencapai akurasi tinggi. Akurasi yang dihasilkan cukup baik, tetapi loss lebih tinggi dibanding ReLU.

Sedangkan model dengan **ReLU** mempercepat proses konvergensi karena menghasilkan gradient yang lebih kuat pada nilai positif. Hal ini membuat model lebih cepat belajar pola data, sehingga loss lebih kecil dan akurasi cenderung lebih tinggi. ReLU juga lebih stabil pada dataset numerik seperti Iris.

---


### **Kesimpulan**

ReLU bekerja lebih efektif pada dataset Iris karena proses belajar lebih cepat, loss lebih kecil, dan akurasi lebih tinggi dibanding Sigmoid yang cenderung lambat akibat masalah vanishing gradient.

---